File: app/src/main/java/ru/p3tr0vich/fuel/FragmentChartCost.java
package ru.p3tr0vich.fuel;

import android.content.Context;
import android.database.Cursor;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.CursorLoader;
import android.support.v4.content.Loader;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.github.mikephil.charting.charts.BarChart;
import com.github.mikephil.charting.components.AxisBase;
import com.github.mikephil.charting.components.LimitLine;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.components.YAxis;
import com.github.mikephil.charting.data.BarData;
import com.github.mikephil.charting.data.BarDataSet;
import com.github.mikephil.charting.data.BarEntry;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.formatter.IAxisValueFormatter;
import com.github.mikephil.charting.formatter.IValueFormatter;
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet;
import com.github.mikephil.charting.utils.ViewPortHandler;

import java.util.ArrayList;
import java.util.Calendar;

import ru.p3tr0vich.fuel.helpers.ContentResolverHelper;
import ru.p3tr0vich.fuel.helpers.DatabaseHelper;
import ru.p3tr0vich.fuel.models.ChartCostModel;
import ru.p3tr0vich.fuel.utils.Utils;
import ru.p3tr0vich.fuel.utils.UtilsDate;
import ru.p3tr0vich.fuel.utils.UtilsFormat;

public class FragmentChartCost extends FragmentBase implements
        LoaderManager.LoaderCallbacks<Cursor> {

    private static final int YEARS_CURSOR_LOADER_ID = 0;
    private static final int CHART_CURSOR_LOADER_ID = 1;

    private TabLayout mTabLayout;
    private BarChart mChart;
    private TextView mTextNoRecords;
    private TextView mTextMedian;
    private TextView mTextSum;

    private final int[] mColors = new int[]{R.color.chart_winter, R.color.chart_winter,
            R.color.chart_spring, R.color.chart_spring, R.color.chart_spring,
            R.color.chart_summer, R.color.chart_summer, R.color.chart_summer,
            R.color.chart_autumn, R.color.chart_autumn, R.color.chart_autumn,
            R.color.chart_winter};

    private boolean mUpdateYearsInProcess = true;

    private ChartCostModel mChartCostModel;

    private static class Months {
        public static final int COUNT = 12;

        private final String[] mMonths = new String[COUNT];

        public Months(@NonNull Context context) {
            Calendar calendar = Calendar.getInstance();
            boolean abbrev = Utils.isPhoneInPortrait();

            for (int month = Calendar.JANUARY; month <= Calendar.DECEMBER; month++)
                mMonths[month] = UtilsDate.getMonthName(context, calendar, month, abbrev);
        }

        @NonNull
        public String getMonth(int month) {
            return month >= Calendar.JANUARY && month <= Calendar.DECEMBER ? mMonths[month] : "";
        }
    }

    Months months;

    @Override
    public int getTitleId() {
        return R.string.title_chart_cost;
    }

    @Override
    public int getSubtitleId() {
        return R.string.title_chart_cost_subtitle;
    }

    private int getYearFromPosition(int index) {
        return mChartCostModel.getYears() != null ? mChartCostModel.getYears()[index] : -1;
    }

    private int getPositionForYear(int year) {
        if (mChartCostModel.getYears() == null) return -1;

        for (int i = 0; i < mChartCostModel.getYears().length; i++)
            if (mChartCostModel.getYears()[i] == year)
                return i;
        return -1;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Context context = getContext();

        assert context != null;

        months = new Months(context);

        if (savedInstanceState == null) {
            mChartCostModel = new ChartCostModel();

            getLoaderManager().initLoader(YEARS_CURSOR_LOADER_ID, null, this);
        } else
            mChartCostModel = savedInstanceState.getParcelable(ChartCostModel.NAME);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(ChartCostModel.NAME, mChartCostModel);
    }

    @Override
    public void onDestroy() {
        mHandler.removeCallbacks(mRunnableShowNoRecords);
        super.onDestroy();
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_chart_cost, container, false);

        mTabLayout = view.findViewById(R.id.tab_layout);

        mTextNoRecords = view.findViewById(R.id.text_no_records);
        mTextNoRecords.setVisibility(View.GONE);

        mTextMedian = view.findViewById(R.id.text_median);
        mTextSum = view.findViewById(R.id.text_sum);

        mChart = view.findViewById(R.id.chart);

        mChart.getDescription().setText("");

        mChart.setNoDataText("");

        mChart.setDrawBarShadow(false);
        mChart.setTouchEnabled(false);
        mChart.setScaleEnabled(false);
        mChart.setDoubleTapToZoomEnabled(false);
        mChart.setPinchZoom(false);
        mChart.setDrawGridBackground(false);
        mChart.setDrawValueAboveBar(true);

        XAxis xAxis = mChart.getXAxis();
        xAxis.setPosition(XAxis.XAxisPosition.BOTTOM);
        xAxis.setDrawAxisLine(true);
        xAxis.setAxisLineWidth(1f);
        xAxis.setDrawGridLines(false);
        xAxis.setTextSize(8f);
        xAxis.setGranularity(1f);
        xAxis.setLabelCount(Months.COUNT);
        xAxis.setAxisMinimum(-0.5f);
        xAxis.setAxisMaximum(Months.COUNT - 0.5f);
        xAxis.setValueFormatter(mMonthFormatter);
        xAxis.setDrawLimitLinesBehindData(true);

        YAxis yAxis = mChart.getAxisLeft();
        yAxis.setEnabled(true);
        xAxis.setDrawAxisLine(true);
        xAxis.setDrawGridLines(false);
        yAxis.setAxisMinimum(-1f);
        yAxis.setDrawLabels(false);
        yAxis.setDrawLimitLinesBehindData(true);

        mChart.getAxisRight().setEnabled(false);
        mChart.getLegend().setEnabled(false);

        mChart.setHardwareAccelerationEnabled(true);

        if (savedInstanceState != null) {
            updateYears();
            updateChart();
        }

        mTabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
                if (!mUpdateYearsInProcess) {
                    setYear(getYearFromPosition(tab.getPosition()));
                }
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });

        view.setOnTouchListener(mOnSwipeTouchListener);
        mChart.setOnTouchListener(mOnSwipeTouchListener);

        return view;
    }

    private void selectTab(@Nullable TabLayout.Tab tab) {
        if (tab != null) {
            tab.select();
        }
    }

    private final OnSwipeTouchListener mOnSwipeTouchListener = new OnSwipeTouchListener(getActivity()) {
        @Override
        public void onSwipeRight() {
            int position = mTabLayout.getSelectedTabPosition();
            if (position != 0) {
                selectTab(mTabLayout.getTabAt(position - 1));
            }
        }

        @Override
        public void onSwipeLeft() {
            int position = mTabLayout.getSelectedTabPosition();
            if (position != mTabLayout.getTabCount() - 1) {
                selectTab(mTabLayout.getTabAt(position + 1));
            }
        }

        @Override
        public void onSwipeTop() {
        }

        @Override
        public void onSwipeBottom() {
        }
    };

    private void updateYears() {
        mUpdateYearsInProcess = true;
        try {
            if (mChartCostModel.getYears() == null || mChartCostModel.getYears().length == 0)
                return;

            for (int year : mChartCostModel.getYears()) {
                mTabLayout.addTab(mTabLayout.newTab().setText(String.valueOf(year)));
            }

            int position = getPositionForYear(mChartCostModel.getYear());
            if (position == -1) {
                position = mChartCostModel.getYears().length - 1;
                mChartCostModel.setYear(mChartCostModel.getYears()[position]);
            }

            selectTab(mTabLayout.getTabAt(position));
        } finally {
            mUpdateYearsInProcess = false;
        }
    }

    private void setYear(int year) {
        if (year == mChartCostModel.getYear()) return;

        mChartCostModel.setYear(year);

        getLoaderManager().restartLoader(CHART_CURSOR_LOADER_ID, null, this);
    }

    private static class ChartCursorLoader extends CursorLoader {

        private final int mYear;

        ChartCursorLoader(Context context, int year) {
            super(context);
            mYear = year;
        }

        @Override
        public Cursor loadInBackground() {
            return ContentResolverHelper.getSumByMonthsForYear(getContext(), mYear);
        }
    }

    private static class YearsCursorLoader extends CursorLoader {

        YearsCursorLoader(Context context) {
            super(context);
        }

        @Override
        public Cursor loadInBackground() {
            return ContentResolverHelper.getYears(getContext());
        }
    }

    @NonNull
    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        switch (id) {
            case CHART_CURSOR_LOADER_ID:
                return new ChartCursorLoader(getContext(), mChartCostModel.getYear());
            case YEARS_CURSOR_LOADER_ID:
                return new YearsCursorLoader(getContext());
            default:
                throw new IllegalArgumentException("Wrong Loader ID");
        }
    }

    @Override
    public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor data) {
        switch (loader.getId()) {
            case CHART_CURSOR_LOADER_ID:
                float sum;
                String month;

                mChartCostModel.clear();
                float[] sums = mChartCostModel.getSums();

                if (data.getCount() > 0) {
                    if (data.moveToFirst()) do {
                        sum = data.getFloat(DatabaseHelper.TableFueling.Columns.COST_SUM_INDEX);
                        month = data.getString(DatabaseHelper.TableFueling.Columns.MONTH_INDEX);

                        sums[Integer.parseInt(month) - 1] = sum;
                    } while (data.moveToNext());

                    mChartCostModel.setSums(sums);
                }

                updateChart();

                break;
            case YEARS_CURSOR_LOADER_ID:
                int count = data.getCount();

                if (count > 0) {
                    int[] years = new int[count];

                    int i = 0;

                    if (data.moveToFirst()) do {
                        years[i] = data.getInt(DatabaseHelper.TableFueling.Columns.YEAR_INDEX);

                        i++;
                    } while (data.moveToNext());

                    mChartCostModel.setYears(years);

                    updateYears();

                    getLoaderManager().initLoader(CHART_CURSOR_LOADER_ID, null, this);
                } else {
                    mChartCostModel.setYears(null);

                    updateChart();
                }
        }
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> loader) {
    }

    private static final class FloatFormatter implements IValueFormatter {

        @Override
        public String getFormattedValue(float value, Entry entry, int dataSetIndex, ViewPortHandler viewPortHandler) {
            return UtilsFormat.floatToString(value, false);
        }
    }

    private final class MonthFormatter implements IAxisValueFormatter {

        @Override
        public String getFormattedValue(float value, AxisBase axis) {
            return months.getMonth((int) value);
        }
    }

    private final FloatFormatter mFloatFormatter = new FloatFormatter();
    private final MonthFormatter mMonthFormatter = new MonthFormatter();

    private float round(float f) {
        return Math.round(f / 10f) * 10f;
    }

    private void updateChart() {
        mHandler.removeCallbacks(mRunnableShowNoRecords);

        if (mChartCostModel.getHasData()) {
            mTextNoRecords.setVisibility(View.GONE);

            ArrayList<BarEntry> sumsEntries = new ArrayList<>();

            float[] sums = mChartCostModel.getSums();
            float sum;
            for (int i = 0; i < sums.length; i++) {
                sum = sums[i];

                sum = round(sum);

                sumsEntries.add(new BarEntry(i, sum));
            }

            BarDataSet sumsSet = new BarDataSet(sumsEntries, "");
            sumsSet.setColors(mColors, getContext());
            sumsSet.setAxisDependency(YAxis.AxisDependency.LEFT);

            ArrayList<IBarDataSet> dataSets = new ArrayList<>();
            dataSets.add(sumsSet);

            BarData sumsData = new BarData(dataSets);
            sumsData.setValueFormatter(mFloatFormatter);
            sumsData.setValueTextSize(8f);
            sumsData.setBarWidth(0.8f);

            mChart.setData(sumsData);

            mChart.getAxisLeft().removeAllLimitLines();

            if (mChartCostModel.getMedian() > 0 && mChartCostModel.isSumsNotEquals()) {
                LimitLine medianLine = new LimitLine(mChartCostModel.getMedian());
                medianLine.setLineColor(Utils.getColor(R.color.chart_median));
                medianLine.setLineWidth(0.5f);
                medianLine.enableDashedLine(8f, 2f, 0f);

                mChart.getAxisLeft().addLimitLine(medianLine);
            }

            int duration = Utils.getInteger(R.integer.animation_chart);
            mChart.animateXY(duration, duration);
        } else {
            mChart.clear();

            mHandler.postDelayed(mRunnableShowNoRecords, Utils.getInteger(R.integer.delayed_time_show_no_records));
        }

        updateTotal();
    }

    private void updateTotal() {
        UtilsFormat.floatToTextView(mTextMedian, round(mChartCostModel.getMedian()), true);
        UtilsFormat.floatToTextView(mTextSum, mChartCostModel.getSum(), true);
    }

    private final Handler mHandler = new Handler();

    private final Runnable mRunnableShowNoRecords = new Runnable() {
        @Override
        public void run() {
            Utils.setViewVisibleAnimate(mTextNoRecords, true);
        }
    };
}

File: app/src/main/kotlin/ru/p3tr0vich/fuel/fragments/FragmentChartCost.kt
package ru.p3tr0vich.fuel.fragments

import android.content.Context
import android.database.Cursor
import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.loader.app.LoaderManager
import androidx.loader.content.CursorLoader
import androidx.loader.content.Loader
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.AxisBase
import com.github.mikephil.charting.components.LimitLine
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.formatter.IAxisValueFormatter
import com.github.mikephil.charting.formatter.IValueFormatter
import com.github.mikephil.charting.interfaces.datasets.IBarDataSet
import com.github.mikephil.charting.utils.ViewPortHandler
import com.google.android.material.tabs.TabLayout
import ru.p3tr0vich.fuel.R
import ru.p3tr0vich.fuel.factories.FragmentFactory
import ru.p3tr0vich.fuel.helpers.ContentResolverHelper
import ru.p3tr0vich.fuel.listeners.OnSwipeTouchListener
import ru.p3tr0vich.fuel.models.ChartCostModel
import ru.p3tr0vich.fuel.models.DatabaseModel
import ru.p3tr0vich.fuel.utils.Utils
import ru.p3tr0vich.fuel.utils.UtilsDate
import ru.p3tr0vich.fuel.utils.UtilsFormat
import java.util.*

class FragmentChartCost : FragmentBase(FragmentFactory.Ids.CHART_COST), LoaderManager.LoaderCallbacks<Cursor> {

    private var tabLayout: TabLayout? = null
    private lateinit var chart: BarChart
    private var textNoRecords: TextView? = null
    private var textMedian: TextView? = null
    private var textSum: TextView? = null

    private val colors = intArrayOf(
            R.color.chart_winter, R.color.chart_winter,
            R.color.chart_spring, R.color.chart_spring, R.color.chart_spring,
            R.color.chart_summer, R.color.chart_summer, R.color.chart_summer,
            R.color.chart_autumn, R.color.chart_autumn, R.color.chart_autumn,
            R.color.chart_winter)

    private var updateYearsInProcess = true

    private var chartCostModel: ChartCostModel? = null

    private lateinit var months: Months

    override val titleId: Int
        get() = R.string.title_chart_cost

    override val subtitleId: Int
        get() = R.string.title_chart_cost_subtitle

    private val onSwipeTouchListener = object : OnSwipeTouchListener(context) {
        override fun onSwipeRight() {
            val position = tabLayout!!.selectedTabPosition

            if (position != 0) {
                selectTab(tabLayout!!.getTabAt(position - 1))
            }
        }

        override fun onSwipeLeft() {
            val position = tabLayout!!.selectedTabPosition

            if (position != tabLayout!!.tabCount - 1) {
                selectTab(tabLayout!!.getTabAt(position + 1))
            }
        }

        override fun onSwipeTop() {}

        override fun onSwipeBottom() {}
    }

    private val floatFormatter = FloatFormatter()
    private val monthFormatter = MonthFormatter()

    private val handler = Handler()

    private val runnableShowNoRecords = Runnable { Utils.setViewVisibleAnimate(textNoRecords!!, true) }

    private class Months(context: Context) {

        private val months = Array(COUNT) { "" }

        init {
            for (month in Calendar.JANUARY..Calendar.DECEMBER)
                months[month] = UtilsDate.getMonthName(context, Calendar.getInstance(), month, Utils.isPhoneInPortrait)
        }

        fun getMonth(month: Int): String {
            return if (month >= Calendar.JANUARY && month <= Calendar.DECEMBER) months[month] else ""
        }

        companion object {
            const val COUNT = 12
        }
    }

    private fun getYearFromPosition(index: Int): Int {
        return chartCostModel!!.years?.get(index) ?: -1
    }

    private fun getPositionForYear(year: Int): Int {
        if (chartCostModel!!.years == null) return -1

        for (i in 0 until chartCostModel!!.years!!.size) {
            if (chartCostModel!!.years!![i] == year) {
                return i
            }
        }

        return -1
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        months = Months(context!!)

        if (savedInstanceState == null) {
            chartCostModel = ChartCostModel()

            loaderManager.initLoader(YEARS_CURSOR_LOADER_ID, null, this)
        } else
            chartCostModel = savedInstanceState.getParcelable(ChartCostModel.NAME)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(ChartCostModel.NAME, chartCostModel)
    }

    override fun onDestroy() {
        handler.removeCallbacks(runnableShowNoRecords)

        super.onDestroy()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {

        val view = inflater.inflate(R.layout.fragment_chart_cost, container, false)

        tabLayout = view.findViewById(R.id.tab_layout)

        textNoRecords = view.findViewById(R.id.text_no_records)
        textNoRecords!!.visibility = View.GONE

        textMedian = view.findViewById(R.id.text_median)
        textSum = view.findViewById(R.id.text_sum)

        chart = view.findViewById(R.id.chart)

        chart.description.text = ""

        chart.setNoDataText("")

        chart.setDrawBarShadow(false)
        chart.setTouchEnabled(false)
        chart.setScaleEnabled(false)
        chart.isDoubleTapToZoomEnabled = false
        chart.setPinchZoom(false)
        chart.setDrawGridBackground(false)
        chart.setDrawValueAboveBar(true)

        with(chart.xAxis) {
            position = XAxis.XAxisPosition.BOTTOM
            setDrawAxisLine(true)
            axisLineWidth = 1f
            setDrawGridLines(false)
            textSize = 8f
            granularity = 1f
            labelCount = Months.COUNT
            axisMinimum = -0.5f
            axisMaximum = Months.COUNT - 0.5f
            valueFormatter = monthFormatter
            setDrawLimitLinesBehindData(true)
        }

        with(chart.axisLeft) {
            isEnabled = true
            setDrawAxisLine(true)
            setDrawGridLines(false)
            axisMinimum = -1f
            setDrawLabels(false)
            setDrawLimitLinesBehindData(true)
        }

        chart.axisRight.isEnabled = false
        chart.legend.isEnabled = false

        chart.setHardwareAccelerationEnabled(true)

        if (savedInstanceState != null) {
            updateYears()
            updateChart()
        }

        tabLayout!!.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
            override fun onTabSelected(tab: TabLayout.Tab) {
                if (!updateYearsInProcess) {
                    setYear(getYearFromPosition(tab.position))
                }
            }

            override fun onTabUnselected(tab: TabLayout.Tab) {}

            override fun onTabReselected(tab: TabLayout.Tab) {}
        })

        view.setOnTouchListener(onSwipeTouchListener)

        return view
    }

    private fun selectTab(tab: TabLayout.Tab?) {
        tab?.select()
    }

    private fun updateYears() {
        updateYearsInProcess = true
        try {
            if (chartCostModel!!.years?.isEmpty() == true) {
                return
            }

            for (year in chartCostModel!!.years!!) {
                tabLayout!!.addTab(tabLayout!!.newTab().setText(year.toString()))
            }

            var position = getPositionForYear(chartCostModel!!.year)
            if (position == -1) {
                position = chartCostModel!!.years!!.size - 1
                chartCostModel!!.year = chartCostModel!!.years!![position]
            }

            selectTab(tabLayout!!.getTabAt(position))
        } finally {
            updateYearsInProcess = false
        }
    }

    private fun setYear(year: Int) {
        if (year == chartCostModel!!.year) return

        chartCostModel!!.year = year

        loaderManager.restartLoader(CHART_CURSOR_LOADER_ID, null, this)
    }

    private class ChartCursorLoader(context: Context, private val year: Int) : CursorLoader(context) {

        override fun loadInBackground(): Cursor? {
            return ContentResolverHelper.getSumByMonthsForYear(context, year)
        }
    }

    private class YearsCursorLoader(context: Context) : CursorLoader(context) {

        override fun loadInBackground(): Cursor? {
            return ContentResolverHelper.getYears(context)
        }
    }

    override fun onCreateLoader(id: Int, args: Bundle?): Loader<Cursor> {
        return when (id) {
            CHART_CURSOR_LOADER_ID -> ChartCursorLoader(context!!, chartCostModel!!.year)
            YEARS_CURSOR_LOADER_ID -> YearsCursorLoader(context!!)
            else -> throw IllegalArgumentException("Wrong Loader ID")
        }
    }

    override fun onLoadFinished(loader: Loader<Cursor>, data: Cursor) {
        when (loader.id) {
            CHART_CURSOR_LOADER_ID -> {
                var sum: Float
                var month: String

                chartCostModel!!.clear()
                val sums = chartCostModel!!.sums

                if (data.count > 0) {
                    if (data.moveToFirst())
                        do {
                            sum = data.getFloat(DatabaseModel.TableFueling.Columns.COST_SUM_INDEX)
                            month = data.getString(DatabaseModel.TableFueling.Columns.MONTH_INDEX)

                            sums[Integer.parseInt(month) - 1] = sum
                        } while (data.moveToNext())

                    chartCostModel!!.sums = sums
                }

                updateChart()
            }
            YEARS_CURSOR_LOADER_ID -> {
                val count = data.count

                if (count > 0) {
                    val years = IntArray(count)

                    var i = 0

                    if (data.moveToFirst())
                        do {
                            years[i] = data.getInt(DatabaseModel.TableFueling.Columns.YEAR_INDEX)

                            i++
                        } while (data.moveToNext())

                    chartCostModel!!.years = years

                    updateYears()

                    loaderManager.initLoader(CHART_CURSOR_LOADER_ID, null, this)
                } else {
                    chartCostModel!!.years = null

                    updateChart()
                }
            }
        }
    }

    override fun onLoaderReset(loader: Loader<Cursor>) {}

    private class FloatFormatter : IValueFormatter {

        override fun getFormattedValue(value: Float, entry: Entry, dataSetIndex: Int, viewPortHandler: ViewPortHandler): String {
            return UtilsFormat.floatToString(value, false)
        }
    }

    private inner class MonthFormatter : IAxisValueFormatter {

        override fun getFormattedValue(value: Float, axis: AxisBase): String {
            return months.getMonth(value.toInt())
        }
    }

    private fun round(f: Float): Float {
        return Math.round(f / 10f) * 10f
    }

    private fun updateChart() {
        handler.removeCallbacks(runnableShowNoRecords)

        if (chartCostModel!!.hasData) {
            textNoRecords!!.visibility = View.GONE

            val sumsEntries = ArrayList<BarEntry>()

            val sums = chartCostModel!!.sums

            var sum: Float

            for (i in sums.indices) {
                sum = sums[i]

                sum = round(sum)

                sumsEntries.add(BarEntry(i.toFloat(), sum))
            }

            val sumsSet = BarDataSet(sumsEntries, "")
            sumsSet.setColors(colors, context)
            sumsSet.axisDependency = YAxis.AxisDependency.LEFT

            val dataSets = ArrayList<IBarDataSet>()
            dataSets.add(sumsSet)

            with(BarData(dataSets)) {
                setValueFormatter(floatFormatter)
                setValueTextSize(8f)
                barWidth = 0.8f

                chart.data = this
            }

            chart.axisLeft.removeAllLimitLines()

            if (chartCostModel!!.median > 0 && chartCostModel!!.isSumsNotEquals) {
                with(LimitLine(chartCostModel!!.median)) {
                    lineColor = Utils.getColor(R.color.chart_median)
                    lineWidth = 0.5f
                    enableDashedLine(8f, 2f, 0f)

                    chart.axisLeft.addLimitLine(this)
                }
            }

            val duration = Utils.getInteger(R.integer.animation_chart)
            chart.animateXY(duration, duration)
        } else {
            chart.clear()

            handler.postDelayed(runnableShowNoRecords, Utils.getInteger(R.integer.delayed_time_show_no_records).toLong())
        }

        updateTotal()
    }

    private fun updateTotal() {
        UtilsFormat.floatToTextView(textMedian, round(chartCostModel!!.median), true)
        UtilsFormat.floatToTextView(textSum, round(chartCostModel!!.sum), true)
    }

    companion object {

        private const val YEARS_CURSOR_LOADER_ID = 0
        private const val CHART_CURSOR_LOADER_ID = 1
    }
}